### ISSUE TITLE
[WebGL and Canvas Rendering] Font does not scale correctly with DPR higher than 1

### ISSUE THREAD

**Issue by CrimsonFez** [xtermjs/xterm.js#4728]
When using WebGL or Canvas addons and displaying on a device with a DPR higher than 1 the text inside the container becomes larger than it should be.

DPR 4
![image](https://github.com/xtermjs/xterm.js/assets/46582994/b415774b-8805-4d10-9f9b-d832459240f8)

DPR 1
![image](https://github.com/xtermjs/xterm.js/assets/46582994/bd2dbb69-1d60-48e2-baf2-0c3f349a9dd7)

When using WebGL I also get this error in Firefox when I set the DPR higher than 1.

`WebGL warning: drawElementsInstanced: Drawing to a destination rect smaller than the viewport rect. (This warning will only be given once)`

## Details
- Firefox and Brave Desktop with Responsive Design Mode enabled both exhibit this
- OS version: Fedora 38
- xterm.js version: 5.2.1
- Canvas addon version: 0.4.0
- WebGL addon version: 0.15.0
- I'm using VueJS 3 to build my SPA

### Steps to reproduce

1. Create a site that uses xtermjs with either canvas or webgl
2. Test the site with Responsive Design Mode

**Comment by Tyriar** [comment]
DPR 4 means 400% zoom so it looks like it's behaving as expected to me? Am I misunderstanding something?

**Comment by CrimsonFez** [comment]
I would expect that the font sizing in the terminal would be the same as the font sizing of regular elements. 
From what I understand about DPR, currently this is unusable on a mobile device.

I currently don't have the ability to test my site on a real mobile device. But I would expect that Responsive Design Mode does a good job of emulating a mobile device screen size.

**Comment by CrimsonFez** [comment]
I might be wrong but, it seams to me like the browser is adjusting the DPR of the canvas before xterm does. Then, xterm is doing its own DPR adjustment which would cause a much larger zoom than intended.

**Comment by Tyriar** [comment]
You're right this is an issue, I think it only applies to the canvas renderer though.

Canvas:

![image](https://github.com/xtermjs/xterm.js/assets/2193314/6e9f73a7-09dd-47f6-8a0f-92140f2eb309)

Webgl:

![image](https://github.com/xtermjs/xterm.js/assets/2193314/9d7ab1ef-0159-45c5-a5bd-c9b75790ed75)

The `WebGL warning: drawElementsInstanced: Drawing to a destination rect smaller than the viewport rect. (This warning will only be given once)` warning only happens once when enabling the addon, it works fine when switching DPR after the fact. Created https://github.com/xtermjs/xterm.js/issues/4731 to track that.

**Comment by jerch** [comment]
@Tyriar I get that behavior on both canvas and webgl. Only the DOM renderer does not show it here. I am not 100% not sure, if thats just an issue of the responsive view in devtools, or if there is really something off with dimension calc (still dont think the latter is the case, as we prolly would see much issue reports then).

**Comment by Tyriar** [comment]
Weird, I also suspect something might be off with the devtools as I've never seen a report of this in practice.

**Comment by jerch** [comment]
Also its stable across browsers, Firefox shows the same weird behavior. :exploding_head:

**Comment by szymonkaliski** [comment]
The issue is reproducible in Chrome if you emulate iPhone 12 and toggle WebGL on and off you can see different sizes of the cells:


https://github.com/xtermjs/xterm.js/assets/1040420/66909753-60f6-4a13-b4c1-ac3532f98ef9

The issue originates in `DevicePixelObserver`:

https://github.com/xtermjs/xterm.js/blob/2edc65baaa50f8b99c67f3df6f8d5d93c70a69e6/src/browser/renderer/shared/DevicePixelObserver.ts#L28-L29

In the demo above, `width` and `height` are `1242` and `2516`, if I change the calculation to:

```
const width = entry.contentBoxSize[0].inlineSize * window.devicePixelRatio;
const height = entry.contentBoxSize[0].blockSize * window.devicePixelRatio;
```

I get `1863` `3774` which produces the proper visual sizing, consistent with Canvas/DOM:

https://github.com/xtermjs/xterm.js/assets/1040420/4c269509-7c8a-4d6c-83c5-986748aea076

Another fix that works is just removing these lines completely: https://github.com/xtermjs/xterm.js/blob/2edc65baaa50f8b99c67f3df6f8d5d93c70a69e6/addons/addon-webgl/src/WebglRenderer.ts#L128-L131

The issue doesn't happen on Safari since it doesn't implement the `devicePixelContentBoxSize`.

**Comment by jerch** [comment]
@szymonkaliski Thx for your investigation.

Your finding with
```javascript
contentBoxSize[0].inlineSize * window.devicePixelRatio
// vs.
devicePixelContentBoxSize[0].inlineSize
```
looks promising as a good fix. Is that guaranteed to yield the same numbers?

Since I did not write the DPR stuff in xterm.js, lets hear @Tyriar about it and possible implications.

**Comment by Tyriar** [comment]
The purpose of `observeDecivePixelDimensions` and `devicePixelContentBoxSize` is to get pixel-perfect actual device pixel count for scaling that is not 1.0 DPR. This was inconsistent and impossible to determine before that API exists, even a single pixel different would make the canvas blurry.

If you change it to `entry.contentBoxSize[0].inlineSize * window.devicePixelRatio` I think it would just bring back that long time problem.

**Comment by jerch** [comment]
@Tyriar Ic, so rounding issues will occur again here. Would it be possible to catch the devtools view with a tiny heuristic, something like this (pseudocode):
```
if abs(contentBoxSize * devicePixelRatio - devicePixelContentBoxSize) > smallish_pixel_incertainty:
  # we have a big discrepancy prolly due to devtools messing DPR up
  return round(contentBoxSize * devicePixelRatio)  # or floor/ceil, whatever is better here
return devicePixelContentBoxSize
```

**Comment by jerch** [comment]
@Tyriar To address #5035 I had a look into the DPR stuff and this issue here is indeed related to `devicePixelContentBoxSize`. That metric is specced to return always device pixels and hence still reports the original values even in the devtools device emulation mode not reflecting the DPR change there.

To get the devtools device view working, we just need to place my suggested fix under https://github.com/xtermjs/xterm.js/blob/a5fc111d623dccf009508489726959afb87e1011/src/browser/renderer/shared/DevicePixelObserver.ts#L8
It will keep native DPR working as usual, but will do the correction for the devtools view. This will indeed re-introduce those rounding issues, but only in the devtools view (which is still better than having it not working at all).
